README.md
---
layout: home
title: "untangle - Extract essence of information and store in fractal structures"
image: assets/favimage-472x472.jpg
---

\[click on image to start the interactive model, NOTE: CPU intensive\]  
[![teaser](assets/favimage-472x472.jpg)](https://rockingship.github.io/untangle-media/shrimp.html)  
\[3D wireframe model of a fractal structure containing zero (yellow), 9 inputs (blue) and 9 outputs (green)\]

# untangle (v2)

Extract essence of information and store/contain in fractal structures

### Welcome to the Wonderful World of fractal logic and computing

_This might be a good spot to break the loop_

All our logic operators share a common component.  
Better worded, all our logic operators are derived from a single operator: the fractal switch.

The fractal world is all about inheritance and transformation.  
Binary values are transformations of a shared reference value: null.

Because there is just a single operator and single value, both can be removed from the equation/system.  
What's left is structure in the form of information and connection.

These structures are self-modifying and `untangle` is their containment field.  
`untangle` isolates and extracts the essence of information.

What `untangle` does is:

  - transform logic expressions into fractal structures
  - let structures digest information
  - solve equations by extracting information

An illustration of some of the many things this project uncovered:
With 9 variables and 4 fractal operators it is possible to construct 1.206e13 different structures.  
After removing all transformations and synonyms, 791647 unique structures remain.
binary-operators used in algebra can describe about 0.1% of them.

### Table of contents

  - [Welcome to the Wonderful World of fractal logic and computing](#welcome-to-the-wonderful-world-of-fractal-logic-and-computing)
    - [Table of contents](#table-of-contents)
    - [Time trail](#time-trail)
    - [Duality](#duality)
    - [Time is a dimension](#time-is-a-dimension)
  - [QTF operator](#qtf-operator)
    - [Traditional implementation](#traditional-implementation)
    - [Unified operator](#unified-operator)
  - [Normalisation](#normalisation)
    - [Level-1 normalisation](#level-1-normalisation)
      - [Variable substitution](#variable-substitution)
      - [Semi-final normalised collection](#semi-final-normalised-collection)
      - [Final normalised collection](#final-normalised-collection)
      - [`QnTF` normalisation](#qntf-normalisation)
      - [Symmetric ordering](#symmetric-ordering)
      - [QnTF implementation](#qntf-implementation)
    - [Level-2 normalisation](#level-2-normalisation)
    - [Level-3 normalisation (cascading)](#level-3-normalisation-cascading)
      - [Structure and skin separation](#structure-and-skin-separation)
      - [Structure sizes](#structure-sizes)
  - [Metrics](#metrics)
  - [Normalising2](#normalising2)
  - [Associative index](#associative-index)
  - [Notation](#notation)
  - [Anatomy of a fractal tree file](#anatomy-of-a-fractal-tree-file)
  - [Structure based compare](#structure-based-compare)
  - [Versioned memory](#versioned-memory)
  - [Member depreciation](#member-depreciation)
  - [Database](#database)
  - [Manifest](#manifest)
  - [Requirements](#requirements)
  - [Building and Installation](#building-and-installation)
  - [Versioning](#versioning)
  - [License](#license)

### Time trail

Untangle is a large project developed over many years.  
The previous version (v1.50) is a working small-scale prototype.  
Feedback revealed a fundamental assumption being invalid.  
About choosing a representative structure/expression from a group of synonyms.  
Smaller is not always better and sometimes even makes it worse.

Version 2 is a rewrite and possibly the first large-scale pre-production platform.  
Version 1.50 will be imported in parts and rewritten with the new paradigm in mind.  
Attention goes all over the place in a non-linear fashion.  
To aid future documentation, a trail of "@date" tokens are used as markers.  
Names, descriptions and comments may refer to future explanations.

### Duality

*It's not a bird, it's not a plane, it's a duale value*

Binary values ("0", "1") imply an ordering. Starting with zero and limited by its base (2).

Logical values ("false", "true") imply an absolute reference.

Duale values ("reference", "NOT-reference") are transformations of a shared reference value.

Just like Yin and Yang.  
Both identical in every aspect except they differ as being "white, not-white" or "black, not-black" or anything else depending on your reference.

### Time is a dimension

Time is used to indicate the flow of information through expressions.  
For example with `a+b*c`, the multiply is performed before the addition.  
Time flows from `*` to `+`.  
Concepts that change the direction of time like loops or conditionals do not exist and need to be loop-unrolled or expanded first.

## QTF operator

<!-- @date 2020-03-05 12:58:00 -->

Information can be described by formulas.
Formulas can be rewritten into expressions.
Expressions can be stored in trees.
Trees are structures consisting of nodes.
Nodes are representations of numeric operators.
Numeric operators can be rewritten into machine instructions.
Numeric machine instructions are networks of AND/OR/XOR gates.
Gates are implemented in silicone.

Nodes can be evaluated by passing electrons through different types of gates.
Information can be extracted from trees by evaluating the nodes.

Information can be rewritten into terms of AND/OR/XOR.
Information is stored as structure (trees) and data (choice of node).

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

TODO: no sorting when read-only
TODO: genpattern save/load
TODO: genpattern list

## [Unreleased]

```
2022-02-20 01:26:44 Fixed: Handling invert in `groupTree_t::addNode()`.
2022-02-17 13:43:00 Changed: `ImportFold()` and `importActive()` do an implicit `rewind()`.
2022-02-17 12:59:19 Deleted: `baseTree_t::system and `grouptree_t::system`.
2022-02-17 12:53:49 Added: `gsave` and `gsystem`. 
2022-02-14 00:30:38 Added: `gvalidate`.
2022-02-13 02:37:35 Changed: Fix `bfold`/`bjoin`/`bload`/`bsystem`/`bvalidate`.
2022-02-10 20:46:09 Changed: Renamed `validate` to `bvalidate`.
2022-02-10 20:44:20 Changed: Modernised `kextract` and renamed to `bextract`.
2022-02-10 18:42:23 Changed: Renamed `eval`/`kload`/`ksave`/`ksystem`.
2022-02-10 18:37:04 Changed: Modernised `entryNames`/`rootNames`/`system`.
2022-02-08 22:14:56 Changed: Improved file/string for `baseTree_t` and `groupTree_t` arguments.
2022-02-08 16:26:58 Changed: Modernised system builders with entrypoints/roots.
2022-02-08 13:56:41 Changed: All system builders (`build*`) export json only, use a loader to convert to a tree.
2022-02-08 01:27:56 Added: `context_t::MAGICMASK_SYSTEM` and upgraded flags in json to text.
2022-02-07 22:55:25 Changed `geval` supports files as input.
2022-02-07 17:39:08 Changed: Allow multi-rooted `groupTree_t` trees as strings.
2022-02-07 22:11:44 Fixed: Code indenting in `groupTree_t::addNode()`.
2022-02-07 10:52:07 Changed: Renamed `key` to `entry` (no code change).
2022-02-07 10:47:27 Fixed: Typos.
2022-02-06 20:30:31 Fixed: Changes for next dataset.
2022-01-31 22:14:23 Changed: Let `groupTree_t::addbasicNode()` determine its own sid.
2022-01-31 20:06:43 Fixed: CRC calculation in `geval` and sync with `eval`/`beval`.
2022-01-31 16:54:35 Changed: Implement `gfold.cc`.
2022-01-31 16:36:15 Fixed: Prefixes in transform strings.
2022-01-31 15:31:50 Fixed: Harden `groupTree_t::loadStringSafe()` against inverted results.
2022-01-31 15:30:26 Added: `groupTree_t::useExpandMember`.
2022-01-31 15:03:16 Changed: Lost+Found
2022-01-31 00:49:07 Added: `groupTree_t::normaliseQTF()` which is a totally revised Q/T/F normalisation method.
2022-01-30 00:14:30 Fixed: Clear node index when `groupTree_t::lookupNode()` fails.
```

## 2022-01-27 21:43:52 [Version 2.13.0]

`groupTree_t`, the engine of untangle, is finally operational. The test wrapper being `geval`.

Some known issues:
 - The ways of self-modifying recursive structures are sometimes weird and unexpected.
 - Sometimes it throws an assert at unexpected moments, even after hours worth of crunch

geval.cc
#pragma GCC optimize ("O0") // usually here from within a debugger

/*
 * geval.cc
 *      Equivalent of `eval.cc` only for `groupTree_t`
 */

/*
 *	This file is part of Untangle, Information in fractal structures.
 *	Copyright (C) 2017-2021, xyzzy@rockingship.org
 *
 *	This program is free software: you can redistribute it and/or modify
 *	it under the terms of the GNU General Public License as published by
 *	the Free Software Foundation, either version 3 of the License, or
 *	(at your option) any later version.
 *
 *	This program is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *	GNU General Public License for more details.
 *
 *	You should have received a copy of the GNU General Public License
 *	along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <ctype.h>
#include <errno.h>
#include <getopt.h>
#include <jansson.h>
#include <stdlib.h>
#include <unistd.h>

#include "context.h"
#include "grouptree.h"
#include "database.h"

/*
 * Resource context.
 * Needs to be global to be accessible by signal handlers.
 *
 * @global {context_t} Application context
 */
context_t ctx;

/**
 * @date 2021-05-17 22:45:37
 *
 * Signal handlers
 *
 * Bump interval timer
 *
 * @param {number} sig - signal (ignored)
 */
void sigalrmHandler(int __attribute__ ((unused)) sig) {
	if (ctx.opt_timer) {
		ctx.tick++;
		alarm(ctx.opt_timer);
	}
}

/**
 * @date 2021-05-13 15:30:14
 *
 * Main program logic as application context
 * It is contained as an independent `struct` so it can be easily included into projects/code
 */
struct gevalContext_t {

	enum {
		/// @constant {number} Size of footprint for `tinyTree_t` in terms of uint64_t
		QUADPERFOOTPRINT = ((1 << MAXSLOTS) / 64)
	};

	/// @var {string} name of database
	const char *opt_databaseName;
	/// @var {number} --datasize, Data vector size containing test patterns for CRC (units in uint64_t)
	unsigned   opt_dataSize;
	/// @var {number} header flags
	uint32_t   opt_flagsSet;
	/// @var {number} header flags
	uint32_t   opt_flagsClr;
	/// @var {number} --maxdepth, Maximum node expansion depth for `groupTree_t`.
	unsigned   opt_maxDepth;
	/// @var {number} --maxnode, Maximum number of nodes for `baseTree_t`.
	unsigned   opt_maxNode;
	/// @var {number} --normalise, display names as normalised with transforms
	unsigned   opt_normalise;
	/// @global {number} --seed=n, Random seed to generate evaluator test pattern
	unsigned   opt_seed;
	/// @var {number} --speed, Speed setting for `groupTree_t`.
	unsigned   opt_speed;

	/// @var {database_t} - Database store to place results
	database_t *pStore;

	gevalContext_t() {
		opt_databaseName = "untangle.db";
		opt_dataSize     = QUADPERFOOTPRINT; // compatible with `footprint_t::QUADPERFOOTPRINT`
		opt_flagsSet     = 0;
		opt_flagsClr     = 0;
		opt_maxDepth     = groupTree_t::DEFAULT_MAXDEPTH;
		opt_maxNode      = groupTree_t::DEFAULT_MAXNODE;
		opt_normalise    = 0;
		opt_seed         = 1;
		opt_speed        = groupTree_t::DEFAULT_SPEED;
		pStore           = NULL;
	}

	/**
	 * @date 2021-06-08 23:45:32
	 *
	 * Calculate the hash of a footprint.
	 *
	 * It doesn't really have to be crc,  as long as the result has some linear distribution over index.
	 * crc32 was chosen because it has a single assembler instruction on x86 platforms.
	 *
	 * Inspired by Mark Adler's software implementation of "crc32c.c -- compute CRC-32C using the Intel crc32 instruction"
	 *
	 * @return {number} - calculate crc
	 */
	uint32_t calccrc32(uint64_t *pData, unsigned numData) const {

		static uint32_t crc32c_table[8][256];

		if (crc32c_table[0][0] == 0) {
			/*
			 * Initialize table
			 */
			uint32_t n, crc, k;
			uint32_t poly = 0x82f63b78;

			for (n = 0; n < 256; n++) {
				crc = n;

				crc = (crc & 1) ? (crc >> 1) ^ poly : (crc >> 1);
				crc = (crc & 1) ? (crc >> 1) ^ poly : (crc >> 1);
				crc = (crc & 1) ? (crc >> 1) ^ poly : (crc >> 1);
				crc = (crc & 1) ? (crc >> 1) ^ poly : (crc >> 1);
				crc = (crc & 1) ? (crc >> 1) ^ poly : (crc >> 1);
				crc = (crc & 1) ? (crc >> 1) ^ poly : (crc >> 1);
				crc = (crc & 1) ? (crc >> 1) ^ poly : (crc >> 1);
				crc = (crc & 1) ? (crc >> 1) ^ poly : (crc >> 1);

				crc32c_table[0][n] = crc;
			}
			for (n = 0; n < 256; n++) {
				crc = crc32c_table[0][n];

				for (k = 1; k < 8; k++) {
					crc = crc32c_table[0][crc & 0xff] ^ (crc >> 8);
					crc32c_table[k][n] = crc;
				}
			}

		}

		/*
		 * Calculate crc
		 */
		uint64_t crc = 0;

		while (numData > 0) {

			crc ^= *pData++;

			crc = crc32c_table[7][crc & 0xff] ^
			      crc32c_table[6][(crc >> 8) & 0xff] ^
			      crc32c_table[5][(crc >> 16) & 0xff] ^
			      crc32c_table[4][(crc >> 24) & 0xff] ^
			      crc32c_table[3][(crc >> 32) & 0xff] ^
			      crc32c_table[2][(crc >> 40) & 0xff] ^
			      crc32c_table[1][(crc >> 48) & 0xff] ^
			      crc32c_table[0][crc >> 56];

			--numData;
		}

		return crc;
	}

	/**
	 * @date 2021-06-08 21:00:46
	 *
	 * Create/load tree based on arguments
	 */
	void handleArgument(const char *inputName) {

		/*
		 * Open input tree
		 */
		groupTree_t *pTree;

		if (strchr(inputName, '.') != NULL) {

			pTree = new groupTree_t(ctx, *pStore);

			/*
			 * Load from file
			 */
			if (pTree->loadFile(inputName)) {
				json_t *jError = json_object();
				json_object_set_new_nocheck(jError, "error", json_string_nocheck("failed to load"));
				json_object_set_new_nocheck(jError, "filename", json_string(inputName));
				ctx.fatal("%s\n", json_dumps(jError, JSON_PRESERVE_ORDER | JSON_COMPACT));
			}

			if (ctx.opt_verbose >= ctx.VERBOSE_VERBOSE) {
				json_t *jResult = json_object();
				json_object_set_new_nocheck(jResult, "filename", json_string_nocheck(inputName));
				pTree->headerInfo(jResult);
				pTree->extraInfo(jResult);
				fprintf(stderr, "%s\n", json_dumps(jResult, JSON_PRESERVE_ORDER | JSON_COMPACT));
				json_delete(jResult);
			}

		} else {
			/*
			 * Load from string
			 */

			pTree = new groupTree_t(ctx, *pStore, inputName, opt_maxNode, /*flags=*/0);

			if (ctx.opt_verbose >= ctx.VERBOSE_VERBOSE) {
				json_t *jResult = json_object();

				jResult = json_object();
				pTree->summaryInfo(jResult);
				fprintf(stderr, "%s\n", json_dumps(jResult, JSON_PRESERVE_ORDER | JSON_COMPACT));
				json_delete(jResult);
			}
		}

		/*
		 * Record footprints for each node to maintain the results to compare trees
		 * Each bit is an independent test.
		 * For ease of calculation, number of tests = number of words per key/node
		 */

		// setup a data vector for evaluation
		footprint_t *pFootprint = (footprint_t *) ctx.myAlloc("pFootprint", pTree->ncount, sizeof(*pFootprint));

		/*
		 * Initialise data/footprint vector
		 */
		if (pTree->nstart - pTree->kstart <= MAXSLOTS) {
			/*
			 * If there are MAXSLOTS keys, then be `eval`/`tinyTree_t` compatible
			 */
			uint32_t kstart = pTree->kstart;
			uint32_t nstart = pTree->nstart;

			uint64_t *v = (uint64_t *) pFootprint;

			// set 64bit slice to zero
			memset(pFootprint, 0, kstart * sizeof(*pFootprint));

			// set footprint for 64bit slice
			assert(MAXSLOTS == 9);
			for (unsigned i = 0; i < (1 << MAXSLOTS); i++) {
				if (kstart + 0 < nstart)
					if (i & (1 << 0)) v[(i / 64) + (kstart + 0) * QUADPERFOOTPRINT] |= 1LL << (i % 64);
				if (kstart + 1 < nstart)
					if (i & (1 << 1)) v[(i / 64) + (kstart + 1) * QUADPERFOOTPRINT] |= 1LL << (i % 64);
				if (kstart + 2 < nstart)
					if (i & (1 << 2)) v[(i / 64) + (kstart + 2) * QUADPERFOOTPRINT] |= 1LL << (i % 64);
				if (kstart + 3 < nstart)
					if (i & (1 << 3)) v[(i / 64) + (kstart + 3) * QUADPERFOOTPRINT] |= 1LL << (i % 64);
				if (kstart + 4 < nstart)
					if (i & (1 << 4)) v[(i / 64) + (kstart + 4) * QUADPERFOOTPRINT] |= 1LL << (i % 64);
				if (kstart + 5 < nstart)
					if (i & (1 << 5)) v[(i / 64) + (kstart + 5) * QUADPERFOOTPRINT] |= 1LL << (i % 64);
				if (kstart + 6 < nstart)
					if (i & (1 << 6)) v[(i / 64) + (kstart + 6) * QUADPERFOOTPRINT] |= 1LL << (i % 64);
				if (kstart + 7 < nstart)
					if (i & (1 << 7)) v[(i / 64) + (kstart + 7) * QUADPERFOOTPRINT] |= 1LL << (i % 64);
				if (kstart + 8 < nstart)
					if (i & (1 << 8)) v[(i / 64) + (kstart + 8) * QUADPERFOOTPRINT] |= 1LL << (i % 64);
			}

		} else {
			srand(opt_seed);

			// fill rest with random patterns
			for (unsigned iEntry = pTree->kstart; iEntry < pTree->nstart; iEntry++) {
				uint64_t *v = (uint64_t *) (pFootprint + iEntry);

				// craptastic random fill
				for (unsigned i = 0; i < opt_dataSize; i++) {
					v[i] = (uint64_t) rand();
					v[i] = (v[i] << 16) ^ (uint64_t) rand();
					v[i] = (v[i] << 16) ^ (uint64_t) rand();
					v[i] = (v[i] << 16) ^ (uint64_t) rand();
				}
			}

			// erase v[0]
			for (unsigned i = 0; i < opt_dataSize; i++)
				pFootprint[0].bits[i] = 0;
		}

		/*
		 * @date 2021-11-03 00:40:18
		 * 
		 * Evaluate test vector (n9) stored in tree.
		 * 
		 * The tree can be accessed from algebra side (Q/T/F)
		 * as well as from the footprint side (2^9=512 bits)
		 * 1- Load QTF structure
		 * 2- Convert to footprint
		 * 3- Lookup with `lookupImprintAssociative()`
		 * 4- Replace with improeved most optimal replacement for most optimal storage
		 * - 4n9-pure (beval) 
		 * geval for groupTree removing structure from the equasion
		 * nodes are expressions of polarised sid/tid, where tid=0
		 * the reduced number of sids 
		 * the cross-poduct. 
		 * from one extreme to Cartesian products of primes
		 * 
		 *  Cartesian product (geval) using grupTree_t
		 *  where the nodes are not what you seek, but what you ask.
		 *  4n9-pure makes is the most minimal dataset covering whole 5n9 addressing space.
		 *  Average of 1.? members per signature group.
		 *  1.?^3 is the expected average of increase of storage.
		 *  
		 * 
		 * maximum storageMost optimal replacement function 
		 * 
		 */

		// load endpoints into placeholder slots
		// load signature into tree
		// apply tree on placeholders
		// `pFootprint` is a replacement for "number", it is a bit vector with the Q/T/F operator
		// the whole tree assumes `tid=0`

		// find group headers
		for (uint32_t iGroup = pTree->nstart; iGroup < pTree->ncount; iGroup++) {
			if (pTree->N[iGroup].gid != iGroup)
				continue; // not a group header

			// top-level components	
			uint32_t Q = 0, Ti = 0, Tu = 0, F = 0;

			// walk through group list in search of a `1n9` node
			for (uint32_t iNode = pTree->N[iGroup].next; iNode != iGroup; iNode = pTree->N[iNode].next) {
				groupNode_t *pNode = pTree->N + iNode;

				// catch `1n9`
				if (pNode->sid == pStore->SID_OR) {
					Q  = pNode->slots[0];
					Ti = IBIT;
					Tu = 0;
					F  = pNode->slots[1];
					break;
				} else if (pNode->sid == pStore->SID_GT) {
					Q  = pNode->slots[0];
					Ti = IBIT;
					Tu = pNode->slots[1];
					F  = 0;
					break;
				} else if (pNode->sid == pStore->SID_NE) {
					Q  = pNode->slots[0];
					Ti = IBIT;
					Tu = pNode->slots[1];
					F  = pNode->slots[1];
					break;
				} else if (pNode->sid == pStore->SID_AND) {
					Q  = pNode->slots[0];
					Ti = 0;
					Tu = pNode->slots[1];
					F  = 0;
					break;
				} else if (pNode->sid == pStore->SID_QNTF) {
					Q  = pNode->slots[0];
					Ti = IBIT;
					Tu = pNode->slots[1];
					F  = pNode->slots[2];
					break;
				} else if (pNode->sid == pStore->SID_QTF) {
					Q  = pNode->slots[0];
					Ti = 0;
					Tu = pNode->slots[1];
					F  = pNode->slots[2];
					break;
				}
			}

			// was anything found
			if (Q == 0) {
				fprintf(stderr, "\n{\"error\":\"group misses 1n9\",\"where\":\"%s:%s:%d\",\"gid\":%u}\n",
					__FUNCTION__, __FILE__, __LINE__, iGroup);
				break;
			}

			// determine if the operator is `QTF` or `QnTF`
			if (Ti) {
				// `QnTF` for each bit in the chunk, apply the operator `"Q ? !T : F"`
				for (unsigned j = 0; j < opt_dataSize; j++)
					pFootprint[iGroup].bits[j] = (pFootprint[Q].bits[j] & ~pFootprint[Tu].bits[j]) ^ (~pFootprint[Q].bits[j] & pFootprint[F].bits[j]);
			} else {
				// `QTF` for each bit in the chunk, apply the operator `"Q ? T : F"`
				for (unsigned j = 0; j < opt_dataSize; j++)
					pFootprint[iGroup].bits[j] = (pFootprint[Q].bits[j] & pFootprint[Tu].bits[j]) ^ (~pFootprint[Q].bits[j] & pFootprint[F].bits[j]);
			}
		}

		uint32_t firstcrc = 0;
		bool     differ   = false;

		for (unsigned iRoot = 0; iRoot < pTree->numRoots; iRoot++) {
			std::string name;
			std::string transform;

			const uint32_t Ri = pTree->roots[iRoot] & IBIT;
			const uint32_t Ru = pTree->roots[iRoot] & ~IBIT;

			// display root name
			printf("%s: ", pTree->rootNames[iRoot].c_str());

			// display footprint
			if (pTree->nstart - pTree->kstart <= MAXSLOTS) {
				// `eval` compatibility, display footprint
				if (Ri) {
					for (unsigned j = 0; j < opt_dataSize; j++)
						printf("%016lx ", pFootprint[Ru].bits[j] ^ ~0U);
				} else {
					for (unsigned j = 0; j < opt_dataSize; j++)
						printf("%016lx ", pFootprint[Ru].bits[j]);
				}
			}

			// display CRC
			unsigned crc32 = calccrc32(pFootprint[Ru].bits, opt_dataSize);
			// Inverted `T` is a concept not present in footprints. As a compromise, invert the result.
			if (Ri)
				crc32 ^= 0xffffffff;
			printf("{%08x} ", crc32);

			if (iRoot == 0)
				firstcrc = crc32;
			else if (firstcrc != crc32)
				differ = true;

			// display expression
			if (opt_normalise) {
				name = pTree->saveString(pTree->roots[iRoot], &transform);
				printf(": %s/%s", name.c_str(), transform.c_str());
			} else {
				name = pTree->saveString(pTree->roots[iRoot]);
				printf(": %s", name.c_str());
			}

			printf("\n");
		}

		if (ctx.opt_verbose >= ctx.VERBOSE_SUMMARY && pTree->estart - pTree->ostart > 1) {
			if (differ)
				fprintf(stderr, "crc DIFFER\n");
			else
				fprintf(stderr, "crc same\n");

			if (differ)
				exit(1);
		}

		ctx.myFree("pFootprint", pFootprint);

		delete pTree;
	}

};

/*
 * Application context.
 * Needs to be global to be accessible by signal handlers.
 *
 * @global {gevalContext_t} Application context
 */
gevalContext_t app;

void usage(char *argv[], bool verbose) {
	fprintf(stderr, "usage: %s <pattern> ...\n", argv[0]);
	if (verbose) {
		fprintf(stderr, "\t-D --database=<filename>   Database to query [default=%s]\n", app.opt_databaseName);
		fprintf(stderr, "\t   --explain\n");
		fprintf(stderr, "\t-n --normalise             Display pattern as: normalised/transform\n");
		fprintf(stderr, "\t-t --numtests=<seconds>    [default=%d]\n", app.opt_dataSize);
		fprintf(stderr, "\t   --maxdeph=<number>      Maximum node expansion depth [default=%d]\n", app.opt_maxDepth);
		fprintf(stderr, "\t   --maxnode=<number>      Maximum tree nodes [default=%d]\n", app.opt_maxNode);
		fprintf(stderr, "\t-q --quiet                 Say less\n");
		fprintf(stderr, "\t   --seed=n                Random seed to generate evaluator test pattern. [Default=%u]\n", app.opt_seed);
		fprintf(stderr, "\t   --speed=<number>        Speed setting [default=%d]\n", app.opt_speed);
		fprintf(stderr, "\t-v --verbose               Say more\n");
		fprintf(stderr, "\t   --timer=<seconds>       [default=%d]\n", ctx.opt_timer);

		fprintf(stderr, "\t   --[no-]paranoid [default=%s]\n", ctx.flags & ctx.MAGICMASK_PARANOID ? "enabled" : "disabled");
		fprintf(stderr, "\t   --[no-]pure [default=%s]\n", ctx.flags & ctx.MAGICMASK_PURE ? "enabled" : "disabled");
		fprintf(stderr, "\t   --[no-]rewrite [default=%s]\n", ctx.flags & ctx.MAGICMASK_REWRITE ? "enabled" : "disabled");
		fprintf(stderr, "\t   --[no-]cascade [default=%s]\n", ctx.flags & ctx.MAGICMASK_CASCADE ? "enabled" : "disabled");
//		fprintf(stderr, "\t   --[no-]shrink [default=%s]\n", ctx.opt_flags &  ctx.MAGICMASK_SHRINK ? "enabled" : "disabled");
//		fprintf(stderr, "\t   --[no-]pivot3 [default=%s]\n", ctx.opt_flags &  ctx.MAGICMASK_PIVOT3 ? "enabled" : "disabled");
	}
}

/**
 * @date 2021-05-13 15:28:31
 *
 * Program main entry point
 * Process all user supplied arguments to construct a application context.
 * Activate application context.
 *
 * @param  {number} argc - number of arguments
 * @param  {string[]} argv - program arguments
 * @return {number} 0 on normal return, non-zero when attention is required
 */
int main(int argc, char *argv[]) {
	setlinebuf(stdout);

	for (;;) {
		enum {
			LO_HELP     = 1, LO_DEBUG, LO_EXPLAIN, LO_MAXDEPTH, LO_MAXNODE, LO_SEED, LO_SPEED, LO_TIMER,
			LO_PARANOID, LO_NOPARANOID, LO_PURE, LO_NOPURE, LO_REWRITE, LO_NOREWRITE, LO_CASCADE, LO_NOCASCADE, LO_SHRINK, LO_NOSHRINK, LO_PIVOT3, LO_NOPIVOT3,
			LO_DATABASE = 'D', LO_DATASIZE = 't', LO_NORMALISE = 'n', LO_QUIET = 'q', LO_VERBOSE = 'v'
		};

		static struct option long_options[] = {
			/* name, has_arg, flag, val */
			{"database",    1, 0, LO_DATABASE},
			{"datasize",    1, 0, LO_DATASIZE},
			{"debug",       1, 0, LO_DEBUG},
			{"explain",     0, 0, LO_EXPLAIN},
			{"help",        0, 0, LO_HELP},
			{"maxdepth",    1, 0, LO_MAXDEPTH},
			{"maxnode",     1, 0, LO_MAXNODE},
			{"normalise",   0, 0, LO_NORMALISE},
			{"quiet",       2, 0, LO_QUIET},
			{"seed",        1, 0, LO_SEED},
			{"speed",       1, 0, LO_SPEED},
			{"timer",       1, 0, LO_TIMER},
			{"verbose",     2, 0, LO_VERBOSE},
			//
			{"paranoid",    0, 0, LO_PARANOID},
			{"no-paranoid", 0, 0, LO_NOPARANOID},
			{"pure",        0, 0, LO_PURE},
			{"no-pure",     0, 0, LO_NOPURE},
			{"rewrite",     0, 0, LO_REWRITE},
			{"no-rewrite",  0, 0, LO_NOREWRITE},
			{"cascade",     0, 0, LO_CASCADE},
			{"no-cascade",  0, 0, LO_NOCASCADE},
//			{"shrink",      0, 0, LO_SHRINK},
//			{"no-shrink",   0, 0, LO_NOSHRINK},
//			{"pivot3",      0, 0, LO_PIVOT3},
//			{"no-pivot3",   0, 0, LO_NOPIVOT3},
			//
			{NULL,          0, 0, 0}
		};

		char optstring[64];
		char *cp                            = optstring;
		int  option_index                   = 0;

		for (int i = 0; long_options[i].name; i++) {
			if (isalpha(long_options[i].val)) {
				*cp++ = (char) long_options[i].val;

				if (long_options[i].has_arg)
					*cp++ = ':';
				if (long_options[i].has_arg == 2)
					*cp++ = ':';
			}
		}

		*cp = '\0';

		int c = getopt_long(argc, argv, optstring, long_options, &option_index);
		if (c == -1)
			break;

		switch (c) {
		case LO_DATABASE:
			app.opt_databaseName = optarg;
			break;
		case LO_DATASIZE:
			app.opt_dataSize = (unsigned) strtoul(optarg, NULL, 10);
			break;
		case LO_DEBUG:
			ctx.opt_debug = (unsigned) strtoul(optarg, NULL, 8); // OCTAL!!
			break;
		case LO_EXPLAIN:
			ctx.opt_debug |= context_t::DEBUGMASK_EXPLAIN;
			break;
		case LO_HELP:
			usage(argv, true);
			exit(0);
		case LO_MAXDEPTH:
			app.opt_maxDepth = (unsigned) strtoul(optarg, NULL, 10);
			break;
		case LO_MAXNODE:
			app.opt_maxNode = (unsigned) strtoul(optarg, NULL, 10);
			break;
		case LO_NORMALISE:
			app.opt_normalise++;
			break;
		case LO_QUIET:
			ctx.opt_verbose = optarg ? (unsigned) strtoul(optarg, NULL, 10) : ctx.opt_verbose - 1;
			break;
		case LO_SEED:
			app.opt_seed = ::strtoul(optarg, NULL, 0);
			break;
		case LO_SPEED:
			app.opt_speed = (unsigned) strtoul(optarg, NULL, 10);
			break;
		case LO_TIMER:
			ctx.opt_timer = (unsigned) strtoul(optarg, NULL, 10);
			break;
		case LO_VERBOSE:
			ctx.opt_verbose = optarg ? (unsigned) strtoul(optarg, NULL, 10) : ctx.opt_verbose + 1;
			break;

		case LO_PARANOID:
			app.opt_flagsSet |= ctx.MAGICMASK_PARANOID;
			app.opt_flagsClr &= ~ctx.MAGICMASK_PARANOID;
			break;
		case LO_NOPARANOID:
			app.opt_flagsSet &= ~ctx.MAGICMASK_PARANOID;
			app.opt_flagsClr |= ctx.MAGICMASK_PARANOID;
			break;
		case LO_PURE:
			app.opt_flagsSet |= ctx.MAGICMASK_PURE;
			app.opt_flagsClr &= ~ctx.MAGICMASK_PURE;
			break;
		case LO_NOPURE:
			app.opt_flagsSet &= ~ctx.MAGICMASK_PURE;
			app.opt_flagsClr |= ctx.MAGICMASK_PURE;
			break;
		case LO_REWRITE:
			app.opt_flagsSet |= ctx.MAGICMASK_REWRITE;
			app.opt_flagsClr &= ~ctx.MAGICMASK_REWRITE;
			break;
		case LO_NOREWRITE:
			app.opt_flagsSet &= ~ctx.MAGICMASK_REWRITE;
			app.opt_flagsClr |= ctx.MAGICMASK_REWRITE;
			break;
		case LO_CASCADE:
			app.opt_flagsSet |= ctx.MAGICMASK_CASCADE;
			app.opt_flagsClr &= ~ctx.MAGICMASK_CASCADE;
			break;
		case LO_NOCASCADE:
			app.opt_flagsSet &= ~ctx.MAGICMASK_CASCADE;
			app.opt_flagsClr |= ctx.MAGICMASK_CASCADE;
			break;
//			case LO_SHRINK:
//				app.opt_flags |=  ctx.MAGICMASK_SHRINK;
//				break;
//			case LO_NOSHRINK:
//				app.opt_flags &=  ~ctx.MAGICMASK_SHRINK;
//				break;
//			case LO_PIVOT3:
//				app.opt_flags |=  ctx.MAGICMASK_PIVOT3;
//				break;
//			case LO_NOPIVOT3:
//				app.opt_flags &=  ~ctx.MAGICMASK_PIVOT3;
//				break;


		case '?':
			ctx.fatal("Try `%s --help' for more information.\n", argv[0]);
		default:
			ctx.fatal("getopt returned character code %d\n", c);
		}
	}

	if (argc - optind < 1) {
		usage(argv, false);
		exit(1);
	}

	/*
	 * Main
	 */

	// set random seed
	if (app.opt_seed)
		srand(app.opt_seed);
	else
		srand(clock());

	// register timer handler
	if (ctx.opt_timer) {
		signal(SIGALRM, sigalrmHandler);
		::alarm(ctx.opt_timer);
	}

	// Open database
	database_t db(ctx);

	db.open(app.opt_databaseName);
	app.pStore = &db;

	// set flags
	ctx.flags = db.creationFlags;
	ctx.flags |= app.opt_flagsSet;
	ctx.flags &= ~app.opt_flagsClr;

	// display system flags when database was created
	if ((ctx.opt_verbose >= ctx.VERBOSE_VERBOSE) || (ctx.flags && ctx.opt_verbose >= ctx.VERBOSE_SUMMARY))
		fprintf(stderr, "[%s] FLAGS [%s]\n", ctx.timeAsString(), ctx.flagsToText(ctx.flags).c_str());

	if (app.opt_maxNode < db.numSignature) {
		fprintf(stderr, "raising --maxnode to %u\n", db.numSignature);
		app.opt_maxNode = db.numSignature;
	}

	while (optind < argc) {
		app.handleArgument(argv[optind]);
		optind++;
	}

	return 0;
}
